### PR TITLE
*: allow REST in API resource

### PR DESCRIPTION
**Pull request by xiang90**
/cc @pwittrock 

I am still testing it out. But I want to see your early feedback on the direction. Thanks!

### CODE REVIEW

File: cmd/apiregister-gen/generators/parser.go [hunk 1]
```diff
@@ -332,8 +358,8 @@ type SubresourceTags struct {
 	REST        string
 }
 
-// ParseSubresourceTag parses the tags in a "+subresource=" comment into a SubresourceTags struct
-func (b *APIsBuilder) ParseSubresourceTag(c *APIResource, tag string) SubresourceTags {
+// ParseResourceTag parses the tags in a "+resource=" comment into a RresourceTags struct
```
Comment (pwittrock) [inline cmd/apiregister-gen/generators/parser.go]: typo : ResourceTags

File: cmd/apiregister-gen/generators/parser.go [hunk 2]
```diff
@@ -324,6 +331,25 @@ func (b *APIsBuilder) GetNameAndImport(tags SubresourceTags) (string, string) {
...
+type ResourceTags struct {
+	Resource string
+	REST     string
+}
+
+// ParseResourceTag parses the tags in a "+resource=" comment into a ResourceTags struct
+func ParseResourceTag(tag string) ResourceTags {
+	args := strings.Split(tag, ",")
+	rest := ""
+	if len(args) == 2 {
```
Comment (pwittrock) [inline cmd/apiregister-gen/generators/parser.go]: Not a blocker for this PR, but I am thinking we will want to change the format of the tag to be more user friendly.

e.g.

```go
// +resource=name:foo,rest:FooRESTImpl
```

And the same for subsourcee

```go
// +subresource=name:scale,request:Scale,rest:ScaleUniversityREST
```

File: cmd/apiregister-gen/generators/parser.go [hunk 3]
```diff
@@ -94,6 +94,8 @@ type APIResource struct {
 	Kind string
 	// Resource is the resource name - e.g. peachescastles
 	Resource string
+	// REST is the rest.Storage implementation used to handle requests
+	REST string
```
Comment (pwittrock) [inline cmd/apiregister-gen/generators/parser.go]: Update comment to say this is optional

File: cmd/apiregister-gen/generators/versioned_generator.go [hunk 4]
```diff
@@ -69,17 +69,30 @@ var (
...
 		{{ range $api := .Resources -}}
 		{{$api.Group}}{{$api.Kind}}Storage,
-		builders.NewApiResource( // Resource status endpoint
-			{{ $api.Group }}.Internal{{ $api.Kind }}Status,
-			{{.Kind}}SchemeFns{},
-			func() runtime.Object { return &{{ $api.Kind }}{} },     // Register versioned resource
-			func() runtime.Object { return &{{ $api.Kind }}List{} }, // Register versioned resource list
-			&{{ $api.Group }}.{{ $api.Kind }}StatusStrategy{builders.StatusStorageStrategySingleton},
-		),
+		{{if $api.REST }}
```
Comment (pwittrock) [inline cmd/apiregister-gen/generators/versioned_generator.go]: I think you want to guard the one above this

> {{$api.Group}}{{$api.Kind}}Storage

The one you are guarding is just for the `/status` subresource, not the actual resource.

**Comment by pwittrock** [comment]
Generally looks good.  Would you add a new apigroup in the example + write an integration test that exercises the REST impl.  You can look at the existing example tests for how to do this.

**Comment by xiang90** [comment]
> Would you add a new apigroup in the example + write an integration test that exercises the REST impl. You can look at the existing example tests for how to do this.

I will finish them in the next few days! Thanks for review and the quick reply.

File: cmd/apiregister-gen/generators/parser.go (hunk 2 above)
Comment (xiang90) [inline cmd/apiregister-gen/generators/parser.go]: tracked here: https://github.com/kubernetes-incubator/apiserver-builder/issues/14

File: example/pkg/apis/miskatonic/v1beta1/university_types.go [hunk 5]
```diff
@@ -26,7 +26,7 @@ import (
 // +genclient=true
 
 // +k8s:openapi-gen=true
-// +resource=universities
+// +resource=universities,University
```
Comment (pwittrock) [inline example/pkg/apis/miskatonic/v1beta1/university_types.go]: This needs to be UniveristyREST if you want it to call `NewUniversityREST()` in the code generator.

File: cmd/apiregister-gen/generators/versioned_generator.go [hunk 6]
```diff
@@ -57,14 +57,26 @@ func (d *versionedGenerator) Finalize(context *generator.Context, w io.Writer) e
 var VersionedAPITemplate = `
 var (
 	{{ range $api := .Resources -}}
-	{{$api.Group}}{{$api.Kind}}Storage = builders.NewApiResource( //  Resource endpoint
-			{{ $api.Group }}.Internal{{ $api.Kind }},
+
+	{{if $api.REST }}
+		{{$api.Group}}{{$api.Kind}}Storage = builders.NewApiResourceWithStorage( // Resource status endpoint
+			{{ $api.Group }}.Internal{{ $api.Kind }}Status,
```
Comment (pwittrock) [inline cmd/apiregister-gen/generators/versioned_generator.go]: What is the usecase for this?  Why are we continuing to create the status endpoint in this case?

Comment (pwittrock) [inline cmd/apiregister-gen/generators/versioned_generator.go]: Is it correct that the intention here is NOT to store any objects, and just expose some RPC endpoint?  If so, what is it doing, and why isn't it a subresource of a resource that actually has some storage?

If the intention is to store the object, why aren't we using the existing storage impl?  Are you trying to override some piece of the behavior?

If you don't want to store the object, why

File: cmd/apiregister-gen/generators/versioned_generator.go [hunk 7]
```diff
@@ -79,7 +91,9 @@ var (
 		{{ range $subresource := $api.Subresources -}}
 		builders.NewApiResourceWithStorage(
 			{{ $api.Group }}.Internal{{ $subresource.REST }},
+			builders.SchemeFnsSingleton,
 			func() runtime.Object { return &{{ $subresource.Request }}{} }, // Register versioned resource
+			nil,
 			&{{ $api.Group }}.{{ $subresource.REST }}{ {{$api.Group}}.New{{$api.Kind}}Registry({{$api.Group}}{{$api.Kind}}Storage) },
```
Comment (pwittrock) [inline cmd/apiregister-gen/generators/versioned_generator.go]: If the resource defines its own REST, we want to follow this pattern, not the storage patterns above.

File: example/pkg/apis/miskatonic/v1beta1/university_types.go [hunk 8]
```diff
@@ -26,7 +26,7 @@ import (
 // +genclient=true
 
 // +k8s:openapi-gen=true
-// +resource=universities
```
Comment (pwittrock) [inline example/pkg/apis/miskatonic/v1beta1/university_types.go]: Sorry I was unclear on slack.  Please create a new type under miskatonic/v1beta1 in a new `yourtype_types.go` file with a new `yourtype_types_test.go` test.

File: cmd/apiregister-gen/generators/versioned_generator.go (hunk 6 above)
Comment (xiang90) [inline cmd/apiregister-gen/generators/versioned_generator.go]: > If the intention is to store the object, why aren't we using the existing storage impl? Are you trying to override some piece of the behavior?

I want to store the objects differently than what today's REST/storage layer does.

**Pull request closed without merging**